presidio_evaluator/evaluation/span_evaluator.py

```diff
@@ -0,0 +1,390 @@
+            prediction_spans = self.merge_adjacent_spans(prediction_spans, sentence_df)
+            
+            # Track matched spans to avoid double-counting
+            matched_preds = set()
+
+            # For each annotation, find best matching prediction
+            for ann_span in annotation_spans:
+                best_match = None
+                best_iou = 0.0
+
+                for pred_span in prediction_spans:
+                    pred_span_key = (
+                        pred_span.entity_type,
+                        pred_span.start_position,
+                        pred_span.end_position
+                    )
+                    if pred_span_key in matched_preds:
+                        continue
+
+                    # Only compare spans of same type
+                    if pred_span.entity_type != ann_span.entity_type:
+                        continue
+
+                    iou = self.calculate_iou(ann_span, pred_span, sentence_df)
+                    if iou > best_iou:
+                        best_iou = iou
+                        best_match = pred_span
+
+                entity_type = ann_span.entity_type
+
+                # If match found above threshold
+                if best_match and best_iou >= self.iou_threshold:
+                    total_true_positives += 1
+                    per_type_metrics[entity_type]["tp"] += 1
+                    matched_preds.add((
+                        best_match.entity_type,
+                        best_match.start_position,
+                        best_match.end_position
+                    ))
+                else:
+                    total_false_negatives += 1
+                    per_type_metrics[entity_type]["fn"] += 1
+                    if best_match:
+                        error_analysis[f"low_iou_{entity_type}"] += 1
+                    else:
+                        error_analysis[f"missed_{entity_type}"] += 1
+
+        # Count false positives (unmatched predictions)
+        for pred_span in prediction_spans:
+            pred_span_key = (
+                pred_span.entity_type,
+                pred_span.start_position,
+                pred_span.end_position
+            )
+            if pred_span_key not in matched_preds:
+                total_false_positives += 1
+                per_type_metrics[pred_span.entity_type]["fp"] += 1
+                error_analysis[f"extra_{pred_span.entity_type}"] += 1
+        
+        # Calculate overall metrics
+        metrics = self._calculate_metrics(
+            total_true_positives,
+            total_false_positives,
+            total_false_negatives
+        )
+        
+        # Calculate per-type metrics
+        per_type = {}
+        for entity_type, counts in per_type_metrics.items():
+            per_type[entity_type] = self._calculate_metrics(
+                counts["tp"], counts["fp"], counts["fn"]
+            )
+        
+        return {
```

Comment:
Consider wrapping the return object in a class for simplifying the interface. If we could reuse the classes used in the old one, that's great, if not (which is likely the case), then let's create new ones.